app/controllers/asset_groups_controller.rb

```diff
@@ -44,37 +40,36 @@ def print
 
   private
 
-    def update_barcodes
-      perform_assets_update
-
-      if @alerts
-        render json: { errors: @alerts }
-      end
-    end
+  def update_barcodes
+    perform_assets_update
 
-    def set_activity
-      # I need the activity to be able to know the step_types compatible to show.
-      @activity = Activity.find(params[:activity_id]) if params[:activity_id]
+    if @alerts
+      render json: { errors: @alerts }
     end
+  end
 
-    # Use callbacks to share common setup or constraints between actions.
-    def set_asset_group
-      @asset_group = AssetGroup.find(params[:id])
-    end
+  def set_activity
+    # I need the activity to be able to know the step_types compatible to show.
+    @activity = Activity.find(params[:activity_id]) if params[:activity_id]
+  end
 
-    def perform_assets_update
-      @asset_group.update_attributes(assets: params_update_asset_group[:assets].map do |uuid_or_barcode|
-                                       Asset.find_or_import_asset_with_barcode(uuid_or_barcode)
-                                     end.compact.uniq)
-    end
+  # Use callbacks to share common setup or constraints between actions.
+  def set_asset_group
+    @asset_group = AssetGroup.find(params[:id])
+  end
 
-    def show_alert(data)
-      @alerts = [] unless @alerts
-      @alerts.push(data)
-    end
+  def perform_assets_update
+    @asset_group.update_attributes(assets: params_update_asset_group[:assets].filter_map do |uuid_or_barcode|
```

Comment:
filter_map is mostly a rubocop suggestion. It is roughly equivalent of map.compact.